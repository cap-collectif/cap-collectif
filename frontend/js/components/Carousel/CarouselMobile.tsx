import React, { PureComponent } from 'react'
import { FormattedDate, FormattedMessage } from 'react-intl'
import moment from 'moment'
import { DatesInterval } from '../Utils/DatesInterval'
import DarkenGradientMedia from '../Ui/Medias/DarkenGradientMedia'
import RatioMediaContainer from '../Ui/Medias/RatioMediaContainer'

type Props = {
  highlighteds: Array<Record<string, any>>
}
type State = {
  translateX: number
  windowWidth: number
}
export class CarouselMobile extends PureComponent<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      translateX: 0,
      windowWidth: window.innerWidth,
    }
  }

  componentDidMount() {
    const carousel = document.querySelector('.carousel__mobile')
    let isDown = false
    let startX

    if (carousel) {
      // touch
      carousel.addEventListener('touchstart', (e: TouchEvent) => {
        isDown = true

        const touchPosition = e.touches && e.touches[0].pageX
        // @ts-expect-error
        startX = touchPosition - carousel.offsetLeft
      })
      // mouse
      carousel.addEventListener('mousedown', (e: MouseEvent) => {
        isDown = true // @ts-expect-error
        startX = e.pageX - carousel.offsetLeft
      })
      carousel.addEventListener('touchleave', () => {
        isDown = false
      })
      carousel.addEventListener('mouseleave', () => {
        isDown = false
      })
      carousel.addEventListener('touchend', () => {
        isDown = false
      })
      carousel.addEventListener('mouseup', () => {
        isDown = false
      })
      carousel.addEventListener('touchmove', (e: TouchEvent) => {
        // @ts-expect-error
        const x = e.touches[0].pageX - carousel.offsetLeft
        this.getTranslateOnMove(x, e, isDown, startX, carousel)
      })
      carousel.addEventListener('mousemove', (e: MouseEvent) => {
        // @ts-expect-error
        const x = e.pageX - carousel.offsetLeft
        this.getTranslateOnMove(x, e, isDown, startX, carousel)
      })
    }

    window.addEventListener('resize', () => {
      this.setState({
        windowWidth: window.innerWidth,
      })
    })
  }

  getTranslateOnMove = (x: number, e: Event, isDown: boolean, startX: number, carousel: Record<string, any>) => {
    const { highlighteds } = this.props
    const { translateX } = this.state
    const carouselItem = carousel && carousel.getElementsByClassName('item')
    const carouselTotalWidth =
      carouselItem[0].offsetWidth * // carousel item width
        (highlighteds.length - 1) - // number of item without visible item
      10
    // remove padding for last item
    if (!isDown) return
    e.preventDefault()
    const walk = (x - startX) / 10

    if (walk + translateX <= 0 && walk + translateX >= -carouselTotalWidth) {
      this.setState({
        translateX: walk + translateX,
      })
    }

    if (walk + translateX > 0) {
      this.setState({
        translateX: 0,
      })
    }

    if (walk + translateX < -carouselTotalWidth) {
      this.setState({
        translateX: -carouselTotalWidth,
      })
    }
  }

  render() {
    const { highlighteds } = this.props
    const { translateX, windowWidth } = this.state
    const translation = `translateX(${translateX}px)`
    return (
      <div className="carousel__mobile">
        <div
          className="mobile__content"
          id="mobile-content"
          style={{
            transform: translation,
          }}
        >
          {highlighteds.map((highlighted, index) => {
            const type = highlighted.object_type
            const item = highlighted[type]
            const maxItemLength = 55
            const trimmedString =
              item.title.length > maxItemLength ? `${item.title.substring(0, maxItemLength)}...` : item.title
            const hadLinearGradient = windowWidth >= 768

            const getMedia = () => {
              if (highlighted[type].media) {
                return <DarkenGradientMedia linearGradient={hadLinearGradient} url={highlighted[type].media.url} />
              }

              if (highlighted[type].cover) {
                return <DarkenGradientMedia linearGradient={hadLinearGradient} url={highlighted[type].cover.url} />
              }

              return <div className="bg--default bg--project" />
            }

            return (
              <div className="item" key={index}>
                <RatioMediaContainer>{getMedia()}</RatioMediaContainer>
                <div className="carousel__content">
                  <p>
                    <span className="carousel__type">
                      <FormattedMessage id={`type-${highlighted.object_type}`} />
                    </span>
                    <br />
                    <a
                      className="carousel__title"
                      href={highlighted[type]._links ? highlighted[type]._links.show : '#'}
                    >
                      {trimmedString}
                    </a>
                    <br />
                    <span className="carousel__date">
                      {type === 'event' && item.startAt && item.endAt && (
                        <DatesInterval startAt={item.startAt} endAt={item.endAt} />
                      )}
                      {type === 'project' && item.startAt && (
                        <FormattedDate
                          value={moment(item.startAt).toDate()}
                          day="numeric"
                          month="long"
                          year="numeric"
                        />
                      )}
                      {type === 'idea' && item.createdAt && (
                        <FormattedDate
                          value={moment(item.createdAt).toDate()}
                          day="numeric"
                          month="long"
                          year="numeric"
                        />
                      )}
                      {type === 'post' && item.publishedAt && (
                        <FormattedDate
                          value={moment(item.publishedAt).toDate()}
                          day="numeric"
                          month="long"
                          year="numeric"
                        />
                      )}
                    </span>
                  </p>
                </div>
              </div>
            )
          })}
        </div>
      </div>
    )
  }
}
export default CarouselMobile
