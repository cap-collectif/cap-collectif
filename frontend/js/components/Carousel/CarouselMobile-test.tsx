/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { CarouselMobile } from './CarouselMobile'

describe('<CarouselDesktop />', () => {
  const props = {
    highlighteds: [
      {
        position: 1,
        id: 4,
        object_type: 'post',
        post: {
          title: 'Immobilier',
          publishedAt: '2018-04-09T23:21:06+0200',
          _links: {
            show: 'capco.test/post/immobilier',
          },
        },
      },
      {
        position: 2,
        id: 3,
        object_type: 'project',
        project: {
          title: 'Croissance, Innovation, Disruption',
          startAt: '2018-04-09T23:21:06+0200',
          _links: {
            show: 'capco.test/project/croissance',
          },
          cover: {
            url: '/media/croissance.jpg',
          },
        },
      },
      {
        position: 3,
        id: 2,
        object_type: 'event',
        event: {
          title: 'ParisWeb 2016',
          startAt: '2018-04-18T16:24:00+0200',
          endAt: '2018-05-12T12:10:00+0200',
          _links: {
            show: 'capco.test/event/parisweb2016',
          },
        },
      },
      {
        position: 4,
        id: 1,
        object_type: 'idea',
        idea: {
          title: 'My idea',
          createdAt: '2018-04-16T18:35:25+0200',
          _links: {
            show: 'capco.test/idea/myidea',
          },
        },
      },
    ],
  }
  const propsEmptyItem = {
    highlighteds: [
      {
        position: 1,
        id: 4,
        object_type: 'post',
        post: {
          title: 'Immobilier',
          publishedAt: '2018-04-09T23:21:06+0200',
          _links: {
            show: 'capco.test/post/immobilier',
          },
        },
      },
    ],
  }
  it('should render correctly', () => {
    const wrapper = shallow(<CarouselMobile {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('should render correctly negative translation', () => {
    const wrapper = shallow(<CarouselMobile {...propsEmptyItem} />)
    wrapper.setState({
      translateX: -300,
    })
    const mobileContent = wrapper.find('#mobile-content').prop('style')
    expect(mobileContent).toHaveProperty('transform', 'translateX(-300px)')
  })
})
