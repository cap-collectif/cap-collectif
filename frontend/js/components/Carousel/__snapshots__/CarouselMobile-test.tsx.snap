// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarouselDesktop /> should render correctly 1`] = `
<div
  className="carousel__mobile"
>
  <div
    className="mobile__content"
    id="mobile-content"
    style={
      Object {
        "transform": "translateX(0px)",
      }
    }
  >
    <div
      className="item"
      key="0"
    >
      <RatioMediaContainer
        ratioX={16}
        ratioY={9}
      >
        <div
          className="bg--default bg--project"
        />
      </RatioMediaContainer>
      <div
        className="carousel__content"
      >
        <p>
          <span
            className="carousel__type"
          >
            <FormattedMessage
              id="type-post"
              values={Object {}}
            />
          </span>
          <br />
          <a
            className="carousel__title"
            href="capco.test/post/immobilier"
          >
            Immobilier
          </a>
          <br />
          <span
            className="carousel__date"
          >
            <FormattedDate
              day="numeric"
              month="long"
              value={2018-04-09T21:21:06.000Z}
              year="numeric"
            />
          </span>
        </p>
      </div>
    </div>
    <div
      className="item"
      key="1"
    >
      <RatioMediaContainer
        ratioX={16}
        ratioY={9}
      >
        <DarkenGradientMedia
          height="100%"
          linearGradient={true}
          url="/media/croissance.jpg"
          width="100%"
        />
      </RatioMediaContainer>
      <div
        className="carousel__content"
      >
        <p>
          <span
            className="carousel__type"
          >
            <FormattedMessage
              id="type-project"
              values={Object {}}
            />
          </span>
          <br />
          <a
            className="carousel__title"
            href="capco.test/project/croissance"
          >
            Croissance, Innovation, Disruption
          </a>
          <br />
          <span
            className="carousel__date"
          >
            <FormattedDate
              day="numeric"
              month="long"
              value={2018-04-09T21:21:06.000Z}
              year="numeric"
            />
          </span>
        </p>
      </div>
    </div>
    <div
      className="item"
      key="2"
    >
      <RatioMediaContainer
        ratioX={16}
        ratioY={9}
      >
        <div
          className="bg--default bg--project"
        />
      </RatioMediaContainer>
      <div
        className="carousel__content"
      >
        <p>
          <span
            className="carousel__type"
          >
            <FormattedMessage
              id="type-event"
              values={Object {}}
            />
          </span>
          <br />
          <a
            className="carousel__title"
            href="capco.test/event/parisweb2016"
          >
            ParisWeb 2016
          </a>
          <br />
          <span
            className="carousel__date"
          >
            <DatesInterval
              day="numeric"
              endAt="2018-05-12T12:10:00+0200"
              hour="numeric"
              minute="numeric"
              month="long"
              second="numeric"
              showCurrentYear={true}
              startAt="2018-04-18T16:24:00+0200"
              year="numeric"
            />
          </span>
        </p>
      </div>
    </div>
    <div
      className="item"
      key="3"
    >
      <RatioMediaContainer
        ratioX={16}
        ratioY={9}
      >
        <div
          className="bg--default bg--project"
        />
      </RatioMediaContainer>
      <div
        className="carousel__content"
      >
        <p>
          <span
            className="carousel__type"
          >
            <FormattedMessage
              id="type-idea"
              values={Object {}}
            />
          </span>
          <br />
          <a
            className="carousel__title"
            href="capco.test/idea/myidea"
          >
            My idea
          </a>
          <br />
          <span
            className="carousel__date"
          >
            <FormattedDate
              day="numeric"
              month="long"
              value={2018-04-16T16:35:25.000Z}
              year="numeric"
            />
          </span>
        </p>
      </div>
    </div>
  </div>
</div>
`;
